Makefile

```diff
@@ -1,174 +1,297 @@
+MODERN_ROM_NAME := $(FILE_NAME)_modern.gba
+MODERN_OBJ_DIR_NAME := $(BUILD_DIR)/modern
 
-ELF = $(ROM:.gba=.elf)
-MAP = $(ROM:.gba=.map)
+ELF_NAME := $(ROM_NAME:.gba=.elf)
+MAP_NAME := $(ROM_NAME:.gba=.map)
+MODERN_ELF_NAME := $(MODERN_ROM_NAME:.gba=.elf)
+MODERN_MAP_NAME := $(MODERN_ROM_NAME:.gba=.map)
 
+# Pick our active variables
+ifeq ($(MODERN),0)
+  ROM := $(ROM_NAME)
+  OBJ_DIR := $(OBJ_DIR_NAME)
+else
+  ROM := $(MODERN_ROM_NAME)
+  OBJ_DIR := $(MODERN_OBJ_DIR_NAME)
+endif
+ELF := $(ROM:.gba=.elf)
+MAP := $(ROM:.gba=.map)
+SYM := $(ROM:.gba=.sym)
+
+# Commonly used directories
 C_SUBDIR = src
-DATA_C_SUBDIR = src/data
 ASM_SUBDIR = asm
+DATA_SRC_SUBDIR = src/data
 DATA_ASM_SUBDIR = data
-SONG_SUBDIR = sound/songs
-MID_SUBDIR = sound/songs/midi
 
 C_BUILDDIR = $(OBJ_DIR)/$(C_SUBDIR)
 ASM_BUILDDIR = $(OBJ_DIR)/$(ASM_SUBDIR)
 DATA_ASM_BUILDDIR = $(OBJ_DIR)/$(DATA_ASM_SUBDIR)
-SONG_BUILDDIR = $(OBJ_DIR)/$(SONG_SUBDIR)
-MID_BUILDDIR = $(OBJ_DIR)/$(MID_SUBDIR)
-
-ASFLAGS := -mcpu=arm7tdmi --defsym REVISION=$(REVISION) --defsym $(GAME_LANGUAGE)=1
 
-CC1             := tools/agbcc/bin/agbcc
-override CFLAGS += -mthumb-interwork -Wimplicit -Wparentheses -Werror -O2 -g -fhex-asm -f2003-patch -ffix-debug-line
+SHELL := bash -o pipefail
 
-CPPFLAGS := -I tools/agbcc -I tools/agbcc/include -iquote include -nostdinc -DREVISION=$(REVISION) -D$(GAME_LANGUAGE)
-
-LDFLAGS = -Map ../../$(MAP)
+# Set flags for tools
+ASFLAGS := -mcpu=arm7tdmi --defsym REVISION=$(REVISION) --defsym $(GAME_LANGUAGE)=1
 
-LIB := -L ../../tools/agbcc/lib -lc -lgcc
+INCLUDE_DIRS := include
+INCLUDE_CPP_ARGS := $(INCLUDE_DIRS:%=-iquote %)
+INCLUDE_SCANINC_ARGS := $(INCLUDE_DIRS:%=-I %)
+
+O_LEVEL ?= 2
+CPPFLAGS := $(INCLUDE_CPP_ARGS) -Wno-trigraphs -DMODERN=$(MODERN)
+ifeq ($(MODERN),0)
+  CPPFLAGS += -I tools/agbcc/include -I tools/agbcc -nostdinc -iquote include -DREVISION=$(REVISION) -D$(GAME_LANGUAGE)
+  CC1 := tools/agbcc/bin/agbcc$(EXE)
+  override CFLAGS += -mthumb-interwork -Wimplicit -Wparentheses -Werror -O$(O_LEVEL) -fhex-asm -f2003-patch -ffix-debug-line -g
+  LIBPATH := -L $(ROOT_DIR)/tools/agbcc/lib
+  LIB := $(LIBPATH) -lgcc -lc 
+  #LIB := $(LIBPATH) -lgcc -lc -L../../libagbsyscall -lagbsyscall
+else
+  # Note: The makefile must be set up to not call these if modern == 0
+  MODERNCC := $(PREFIX)gcc
+  PATH_MODERNCC := PATH="$(PATH)" $(MODERNCC)
+  CC1 := $(shell $(PATH_MODERNCC) --print-prog-name=cc1) -quiet
+  override CFLAGS += -mthumb -mthumb-interwork -O$(O_LEVEL) -mabi=apcs-gnu -mtune=arm7tdmi -march=armv4t -fno-toplevel-reorder -Wno-pointer-to-int-cast
+  LIBPATH := -L "$(dir $(shell $(PATH_MODERNCC) -mthumb -print-file-name=libgcc.a))" -L "$(dir $(shell $(PATH_MODERNCC) -mthumb -print-file-name=libnosys.a))" -L "$(dir $(shell $(PATH_MODERNCC) -mthumb -print-file-name=libc.a))"
+  # LIB := $(LIBPATH) -lc -lnosys -lgcc -L../../libagbsyscall -lagbsyscall
+endif
+# Enable debug info if set
+ifeq ($(DINFO),1)
+  override CFLAGS += -g
+endif
```

Comment:
From what I see, -g is only missing when building modern. And IMO it should be on by default.

Reply:
Done. What I did was set DINFO to 1 by default either on modern or nonmodern.